Add support for bootsel activity LED (#316)
raspberrypi/picotool: BUILDING.md

In order for the SDK to find `picotool` in this custom folder, you will usually need to set the `picotool_DIR` variable in your project. This can be achieved either by setting the `picotool_DIR` environment variable to `$MY_INSTALL_DIR/picotool`, by passing `-Dpicotool_DIR=$MY_INSTALL_DIR/picotool` to your `cmake` command, or by adding `set(picotool_DIR $MY_INSTALL_DIR/picotool)` to your CMakeLists.txt file.

> See the [find_package documentation](https://cmake.org/cmake/help/latest/command/find_package.html#config-mode-search-procedure) for more details

## Customising the build

The `picotool` build has some CMake variables you can use to customise how the build works, and how `picotool` functions:

> - `GENERATE_FIXED_DOCS_WIDTH`: By default the width of the output from `picotool` adjusts to the width of the terminal. Setting this to `true` fixes the width to 140, which is used when generating the [README](README.md).

Reviewer: It's probably outside the scope of this PR, but I wonder if it might make sense to switch this from being a boolean to being an integer? So that you'd then do `-DGENERATE_FIXED_DOCS_WIDTH=140` rather than `-DGENERATE_FIXED_DOCS_WIDTH=true`?

Author: I think keeping the default in the CMakeLists.txt file is better, so you don't have to look up the number to pass on the command line when you just want to update the readme. But maybe a way to override the default could be added using a different variable in another PR - although I'm not really sure what the use cases would be for that?

Reviewer: Ahh, if it's **only** used when updating the readme, then perhaps the existing behaviour doesn't need changing.

> - `DEFAULT_BOOTSEL_LED`: This can be used to set the default value for the `--led` argument, so `picotool` always reboots devices to BOOTSEL with that LED flashing.

Reviewer: Should there also be a `DEFAULT_BOOTSEL_LED_ACTIVE_LOW` ? :thinking:

Author: I think that's not necessary, especially because it would cause issues with turning it off (would need to invert the behaviour of the `--active-low` argument)

Reviewer: Good point, it hadn't occurred to me that `--active-low` just being a flag, rather than a parameter that takes an option, would make this "tricky" :slightly_smiling_face:

> - `DEFAULT_BOOTSEL_LED`: This can be used to set the default value for the `--bootsel-led` argument, so `picotool` always reboots devices to BOOTSEL with that LED flashing.

Reviewer: Perhaps "with that LED flashing" is a bit of a misnomer, and
```suggestion
- `DEFAULT_BOOTSEL_LED`: This can be used to set the default value for the `--bootsel-led` argument, so `picotool` always reboots devices to BOOTSEL with that LED used as the activity indicator.
```
would be better?

- `PICOTOOL_NO_LIBUSB`: By default `picotool` is compiled with USB support if libusb is found. Setting this to `true` explicitly compiles without USB support, which is used when the Pico SDK builds picotool.

> - `USE_PRECOMPILED`: By default the build uses pre-compiles ELF/BIN files for code that is run on the device (enc_bootloader, xip_ram_perms, and picoboot_flash_id). Setting this to `false` re-compiles these files instead.

Reviewer: ```suggestion
- `USE_PRECOMPILED`: By default the build uses pre-compiled ELF/BIN files for code that is run on the device (enc_bootloader, xip_ram_perms, and picoboot_flash_id). Setting this to `false` re-compiles these files instead.
```

> These can all be set by passing `-DNAME=VALUE` to your `cmake` command (e.g. `-DGENERATE_FIXED_DOCS_WIDTH=1`).

Reviewer: Here you've got `-DGENERATE_FIXED_DOCS_WIDTH=1`, but a few lines earlier you said "`GENERATE_FIXED_DOCS_WIDTH`: ... Setting this to `true` ...".

If this is a case of "CMake treats 1 as true and 0 as false", then perhaps it makes sense to change the earlier phrasing to "Setting this to `1` ..." ?

Author: CMake has lots of values it treats as `true` or `false` (https://cmake.org/cmake/help/latest/command/if.html#constant) - I'll change this example to `true` to match the text above

Author: ```suggestion
These can all be set by passing `-DNAME=VALUE` to your `cmake` command (e.g. `-DGENERATE_FIXED_DOCS_WIDTH=true`).
```

> Note that the behaviour when setting `PICOTOOL_NO_LIBUSB` or `USE_PRECOMPILED` is undefined when using an existing build directory, due to the use of CMake external projects - instead you should create a new build directory to ensure it is in the correct state.

Reviewer: Perhaps this would be clearer as:
```suggestion
Note: when setting `PICOTOOL_NO_LIBUSB` or `USE_PRECOMPILED` you should create a new build directory (rather than using an existing build directory), due to the way `picotool` uses external CMake projects.
```
rather than talking about "undefined behaviour"?
